### ISSUE TITLE
difficulties with multiple commands in the shell

### ISSUE THREAD

**Issue by devmanbr** [mscdex/ssh2#1321]
a question that has been really bugging me for the last week is how to make use of multiple shell commands properly using this library.

the only effective approach that I managed to get fully working was putting all the SSH2 code inside a promise, and creating one for each command to be executed (yes, redoing the connection for each command, which is not ideal at all).

thing seemed simpler using the `exec` method, but the need for the interactive shell and its effects forces me to use the `shell` method.

one of the biggest problems I face is that I have to get the output of each command, and only then proceed to the next one.

if the library could be simplified in this sense, or someone can come up with a solution (even a temporary one), I would be very grateful!

**Comment by devmanbr** [comment]
I understand. unfortunately it has to be using `shell()`.

I did a lot of searching on the internet, and saw a lot of comments in various places that hinted to me that they found solutions, but there was no code. I also found the `ssh2shell` package, which promises to solve the problem, but it doesn't have any typing and I didn't understand anything that was done in CoffeeScript.

it's a problem I know a lot of people face, and possibly some more experienced dev managed to solve it effectively (or can do it quickly). If anyone can offer this here, I would be immensely grateful!

**Comment by axkibe** [comment]
"redoing the connection for each command, which is not ideal at all"

can't use ssh sessions for this? one session per command, but all sessions are in one connection.

other than that consider having an agent (something you coded) run on the other side and give you the results.

**Comment by devmanbr** [comment]
is the connection not dropped whenever the `exit` command is sent to shell, or have I misunderstood?

**Comment by axkibe** [comment]
See the ControlMaster option from ssh itself (-M), it can have a stay alive conection with multiple ssh calls.

**Comment by mscdex** [comment]
> is the connection not dropped whenever the `exit` command is sent to shell, or have I misunderstood?

If you do a `ssh.shell()` and you write `'exit\n'` to it one way or another, all it does is terminate that interactive shell session. You're free to continue using the underlying ssh connection to do whatever you want. The one thing you will need to keep in mind though is that some servers have various limits in place, including how many open "sessions" (this includes both `exec()` and `shell()` IIRC) you can have per connection. OpenSSH defaults this limit to 10 the last time I saw. If you try to open more than the server will allow, you'll just get an error passed to your callback.

**Comment by ndraiman** [comment]
I encountered the same issue with running multiple commands in a single SSH connection.
In version `1.11.0`, running multiple commands one after the other worked fine.

Upgrading to anything after `1.11.0` causes one of the commands to fails with error:
> (SSH) Channel open failure: open failed

One solution would be to remain with version `1.11.0`.
But I found that adding a delay of 250ms between each command solved the issue for me.

Hope this helps.
